src/sphmultipcqueue.cpp

```diff
@@ -0,0 +1,599 @@
+/*
+ * Copyright (c) 2016 IBM Corporation.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ * 
+ * Contributors:
+ *     IBM Corporation, Steven Munroe - initial API and implementation
+ *     IBM Corporation, Paul Clarke   - MPMC implementation
+ */
+
+//#define __SASDebugPrint__ 1
+
+#define sas_printf printf
+#include <stdlib.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <time.h>
+#ifdef __HTM__
+#include <htmxlintrin.h>
+#endif
+#include "sasalloc.h"
+#include "freenode.h"
+#ifdef __SASDebugPrint__
+#include "sasio.h"
+#endif
+#include "sasanchr.h"
+#include "sasatom.h"
+#include "sassim.h"
+#include "saslock.h"
+#define SPH_INTERNAL
+#include "sphmultipcqueue.h"
+#include "sphthread.h"
+
+#ifdef __SASDebugPrint__
+#define debug_printf(...) sas_printf(__VA_ARGS__)
+#else
+#define debug_printf(...)
+#endif
+
+typedef struct SPHMPMCQHeader
+{
+	SASBlockHeader blockHeader;
+	longPtr_t startq;
+	longPtr_t endq;
+	longPtr_t align_mask;
+	unsigned int options;
+	unsigned short default_entry_stride;
+	void *dummy6[15]; // ensure qhead in own cacheline
+	longPtr_t qhead;
+	void * dummy7[15]; // ensure qtail in own cacheline
+	longPtr_t qtail;
+	freeNode *headerFreeSpace;
+} SPHMPMCQHeader;
+
+#ifdef __LP64__
+#define HEAP_OFFSET 128
+#if 0
```

Comment:
That would match sphsinglepcqueue.cpp.  I'll change to match.
This area of code is a bit fragile, in that DEFAULT_PAGE _must_ be larger than sizeof the respective header struct, from a later line:
    remaining = DEFAULT_PAGE - sizeof (SPHMPMCQHeader);
So changes to the header could cause breakage.